x/slashing/keeper/infractions.go

```diff
@@ -36,30 +36,105 @@ func (k Keeper) HandleValidatorSignatureConcurrent(ctx sdk.Context, addr cryptot
 
 	window := k.SignedBlocksWindow(ctx)
 
+	index := signInfo.IndexOffset % k.SignedBlocksWindow(ctx)
+
 	missedInfo, found = k.GetValidatorMissedBlocks(ctx, consAddr)
+
 	if !found {
+		arrLen := window / 64
+		if window%64 > 0 {
+			arrLen += 1
+		}
 		missedInfo = types.ValidatorMissedBlockArray{
-			Address:       consAddr.String(),
-			MissedHeights: make([]int64, 0),
+			Address:      consAddr.String(),
+			WindowSize:   window,
+			MissedBlocks: make([]uint64, arrLen),
 		}
 	}
-	cutoff := height - window
-	// go through and increment number evicted as necessary
-	numberEvicted := 0
-	for _, missedHeight := range missedInfo.MissedHeights {
-		if missedHeight <= cutoff {
-			numberEvicted += 1
-		} else {
-			break
+	// TODO:  resize if necessary?
+	if found && missedInfo.WindowSize != window {
+		// we need to resize the missed block array AND update the signing info accordingly
+		switch {
+		case missedInfo.WindowSize < window:
+			// missed block array too short, lets expand it
+			boolArray := k.ParseBitGroupsToBoolArray(missedInfo.MissedBlocks, missedInfo.WindowSize)
```

Comment:
Why do we parse bits into boolean array instead of just appending more uint64 to the bitsArray? And I think the below logics to copy things are also not needed when expanding the window.